src/OrchardCore/OrchardCore.BackgroundTasks/BackgroundTaskService.cs

```diff
@@ -60,12 +60,18 @@ public void Activate()
         // c.f. http://stackoverflow.com/questions/25007670/using-async-await-inside-the-timer-elapsed-event-handler-within-a-windows-servic
         private async void DoWorkAsync(object group)
         {
-            // DoWork is not re-entrant as Timer will not call the callback until the previous callback has returned.
-            // This way if a tasks takes longer than the period itself, DoWork is not called while it's still running.
+            // DoWork needs to be re-entrant as Timer may call the callback before the previous callback has returned.
+            // So, because a task may take longer than the period itself, DoWork needs to check if it's still running.
             ShellContext shellContext = _orchardHost.GetOrCreateShellContext(_shellSettings);
 
             var groupName = group as string ?? "";
 
+            // Check if any task of this group is not idle
```

Comment:
Remove this block, it's unnecessary as the loop does "exactly" the same thing if you move the `taskName` initialization in the `try/catch`

Reply:
Hmm, here it's to check the whole group to see if there is any task e.g which is still running. But maybe you mean that it is allowed to start executing a group even if not all tasks have been executed.

That said, here i already saw that it is still possible when a task becomes `Idle` and before the next one becomes `Running`, here we would need a state for the whole group. Or maybe put all tasks of a given group to `Idle` only when the last one has completed. But before doing anything, do we want to prevent "group reentrancy"?

Then, in the loop we check individually each task as before to see e.g if `Terminate()` has not been called.

Sorry, i don't see the difference to put `taskName = task.GetType().FullName` in the `try/catch` but i can do it.